### ISSUE TITLE
Suspending resources provides no confirmation that they are suspended

### ISSUE THREAD

**Issue by lindblombr** [fluxcd/flux2#4282]
### Describe the bug

`flux suspend` or setting `spec.suspend: true` on FluxCD resources _does_ have the intended effect of pausing reconciliation, so there's no bug with respect to functionality.  Users of FluxCD, however, would prefer some confirmation that Flux controllers recognize that the resource is in a suspended state rather than relying on the presents of `spec.suspend: true` being the only shared truth value.  Some value in the resource's `status` field, acknowledging resource suspension, so that resource users can have a bit more confidence that the suspension is being honored would be appreciated.

### Steps to reproduce

1. Create any FluxCD resource
2. Suspend the resource
3. Look for any indication in the resource, beyond `spec.suspend: true` for confirmation
4. You will not find any such confirmation

### Expected behavior

1. Create any FluxCD resource
2. Suspend the resource
3. Look for `status.suspended` to also reflect `true`
4. Confirmation achieved

### OS / Distro

N/A

### Flux version

>=v2.0.0

### Flux check

N/A

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by hiddeco** [comment]
I am missing a real argument to favor reflection of a static spec value over simply taking the spec for granted.

The confirmation itself is in the fact that the resource itself will stop being updated, including any updates to the Status.

**Comment by lindblombr** [comment]
> I am missing a real argument to favor reflection of a static spec value over simply taking the spec for granted.

I ran into this assisting someone with a HelmRelease resource.  Perhaps we were mis-using the suspend functionality, but the intention was to suspend the resource, remove it, and leave all of the controller provisioned resources in-tact/un-touched.  The question came up "can we confirm that we're indeed suspended before we go and delete the HelmRelease?".  That answer, was, unfortunately, no, that beyond the `spec.suspend` field, there was no other mechanism to determine suspension status or that Flux would treat the resource as expected.  The nature of the ask is, basically, to get an acknowledgement that both the user and Flux have a mutual understanding that reconciliation should be suspended.  Perhaps in this instance, there is a better way to achieve the desired outcome (remove the HelmRelease, touch none of the HelmRelease-managed resources that are created) of which I am unaware.  For instance, going a step further, we could disable helm-controller entirely during the procedure, though that process may be problematic for larger multi-tenant use-cases.  We would likely, then, clear the finalizer, to see the resource removed.

> The confirmation itself is in the fact that the resource itself will stop being updated, including any updates to the Status.

If that is the full scope intended for suspend functionality then it make sense not to update any status.  I just wonder if, in cases like above, not having that status leaves ambiguity that could be easily addressed.

**Comment by hiddeco** [comment]
The fact that the Kubernetes API server reflects the `.spec.suspend` _is_ your confirmation (by contract).

Adding a reflection of this in the Status would not yield any direct benefits, as trusting this reflection is as (or less) reliable as looking at the spec of the object. Given the controller could in both cases (in case of a bug) still decide to proceed and actually do things to the release.

**Comment by lindblombr** [comment]
@hiddeco This was my thinking, too, when first posed with the question :)  It somehow felt a bit circular, however, to answer "How do we know the resource is suspended?" with "Because we suspended it" in this instance. This made me want to pose the question to a more general audience.  I appreciate your answers.

**Comment by hiddeco** [comment]
You are welcome! 

For the sake of documenting thoughts in full: I think this _may_ be different in case a suspend definition would be more complex.

Take an example where it would depend on a defined cron expression, you can have the desire to reflect it being active in the Status of the object, as it would depend on the interpretation of the definition and runtime environment variables like the current time (which would be much harder to reason about for a user than a simple boolean).

**Comment by sigilioso** [comment]
I came across this issue with a similar use-case: I also need to delete a HelmRelease and leave the provisioned resources untouched. I have a follow-up question concerning the HelmRelease status.

After updating `.spec.suspend=true`, I've observed that the field `.status.observedGeneration` in the HelmRelease is incremented. 

Question: (besides the fact that `.spec.suspend` is effectively updated) would it be reliable to check that the `.status.observedGeneration` is effectively increased to confirm that the helm-controller has processed the suspension request?

**Comment by matheuscscp** [comment]
I wouldn't say it's strictly necessary, as helm-controller will simply no longer touch the HelmRelease after you send a successful request to the Kubernetes API updating the field (via `flux suspend`).

**Comment by sigilioso** [comment]
Thanks for your help!

There is still something I don't fully understand. I've observed the following:

1.- The field is updated (as a result `.spec.suspend` becomes true and `.metadata.resourceVersion` is increased.
2.- Right after, `.spec.suspend` remains true and both `.metadata.resourceVersion` and `.status.observedGeneration` are increased.

```
$ kubectl get helmrelease testing-helm-release -o custom-columns="SUSPEND:.spec.suspend,RESOURCE_VERSION:.metadata.resourceVersion,OBSERVED_GENERATION:.status.observedGeneration" --watch
SUSPEND   RESOURCE_VERSION   OBSERVED_GENERATION
false     11752              12
true      11771              12
true      11772              13
```

Therefore I thought that the field `.status.observedGeneration` was updated after the `.spec.suspend` field in the HelmRelease is updated (I was assuming that it was updated when the helm-controller effectively processes the request). Isn't that right?

**Comment by matheuscscp** [comment]
Yes helm-controller will still patch the `.status` if it sees a suspended object: https://github.com/fluxcd/helm-controller/blob/8555152550f8c62b82ce593ce7b5ee3cd5933c1e/internal/controller/helmrelease_controller.go#L127-L166

When you update anything in the object, the Kubernetes API will always bump `.metadata.resourceVersion`. When you change something specifically in the `.spec`, the Kubernetes API will also bump the `.metadata.generation`, and helm-controller will patch the object `.status.observedGeneration` to match this new `.metadata.generation`, but it won't do anything else.
